sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -513,13 +514,16 @@ class MicroBatchExecution(
 
     val triggerLogicalPlan = sink match {
       case _: Sink => newAttributePlan
-      case s: StreamingWriteSupportProvider =>
-        val writer = s.createStreamingWriteSupport(
-          s"$runId",
-          newAttributePlan.schema,
-          outputMode,
-          new DataSourceOptions(extraOptions.asJava))
-        WriteToDataSourceV2(new MicroBatchWrite(currentBatchId, writer), newAttributePlan)
+      case s: SupportsStreamingWrite =>
+        // TODO: we should translate OutputMode to concrete write actions like truncate, but
+        // the truncate action is being developed in SPARK-26666.
```

Comment:
This also needs to validate that OutputMode is not Update and throw an exception if it is. OutputMode.Update can lead to undefined behavior.

Reply:
Won't this mean that DataSourceV2 writers can't handle the full DataStreamWriter interface?

Reply:
I'm not very familiar with streaming internals so I don't know what's the exact semantic of this UPDATE mode.

I believe we can treat it as `SaveMode`: replace it with the new write operators(Append, Truncate, OverwriteByExpression, etc.) and make the semantic clear.

Anyway this is just a TODO comment, we can have more discussion in the PR that address this TODO.

Reply:
This wasn't addressed. Although my review input was on a comment line, I wasn't saying that this needs to be done when the TODO item is done. The v2 implementation needs to fail when passed a dangerous config setting.

Reply:
I spent more time looking into the streaming output mode. The UPDATE mode is by design and is useful when the watermark is not specified, i.e. users don't want to drop any record no matter how late the record is.

The output mode is applied to the entire streaming query, so all the stateful operators and the sinks need to support it.

The `StateStoreSaveExec` support UPDATE mode with a well-defined semantic: it saves all the input records to state store and output them to downstream operators, so the downstream aggregate/join can produce results earlier. e.g. the first epoch outputs
```
|     date     | count |
|   2018-1-1   |   1   |
|   2018-1-2   |   2   |
```
and the next eppch outputs
```
|     date     | count |
|   2018-1-2   |   1   |
|   2018-1-3   |   3   |
```

However, if there are 2 aggregates, the second aggregate can't support UPDATE mode and deal with the early results from the first aggregate. Fortunately we disallow more than one streaming aggregate in a query, so there is no problem.

The other problem is, all the existing sinks do not support UPDATE mode. They either ignore output mode completely, or treat UPDATE same as APPEND. I think they should fail if output mode is UPDATE, as they don't really support it.

Overall, I think UPDATE mode is useful but is hard to implement. We need a mechanism to propagate the "update key" to down stream sateful operators and the sink. That's why I think no source can support UPDATE mode unless we improve the streaming framework to progate the "update key".

Is it possible to remove UPDATE mode completely and re-add it when we have a detailed design? cc @marmbrus @tdas @jose-torres

Reply:
We cannot just drop a widely used feature.  While we don't support all combinations of operators / output modes, people know what the key is and can use that information when configuring custom sinks.

Reply:
> people know what the key is and can use that information when configuring custom sinks

Ah good point! This is kind of a mechanism to propagate "update key": by the user himself.

To continue to support it, we need to create a new mixin trait for `WriteBuilder` to represent UPDATE mode. We can mark it as unstable. @rdblue do you agree with it?

Reply:
@marmbrus, I'm not trying to drop update as a feature. I'm saying that we need to define how it works in the v2 API. That requires telling a sink what the update key is, as @cloud-fan said. I've already proposed a way to do that.

I don't think it is a good idea to rely on "people know" in the v2 API. Spark needs to have reliable behavior across sinks and not communicating the update key is going to cause undefined behavior.

@cloud-fan: For the idea of using sink configuration to pass the update key: why? What is the value of this beyond using the already proposed solution for Spark to pass the update key to the source using a SupportsUpdate API?

Reply:
Sorry for being unclear. My comment was in response to wenchen's suggestion that we "remove UPDATE mode completely and re-add it when we have a detailed design?".

The feature has utility even without passing the key.  That said, I agree it would be better if the key is available in the V2 API.

Reply:
Okay, thanks for the clarification. I have no problem keeping this as-is in v1. I just want to make sure we fix those problems when we introduce the features in v2. Given that we have a proposal for how to do that, I think we should use it instead of continuing to add code that uses an unreliable mechanism.

Reply:
Perhaps I should say this more strongly.  It would be much better if we pass the key information in a unified way.  We just didn't do it because we didn't want to be changing V1, given the belief that V2 was coming.

Reply:
Removing UPDATE mode was not my suggestion. I just raise it and wait for someone to say no.

> I've already proposed a way to do that.

I thought your proposal was to forbid UPDATE mode in all streaming sinks, and seems I was wrong. To confirm, you mean we should only forbid UPDATE mode in v2 sinks, right? @rdblue

Reply:
@cloud-fan, I don't want to remove the update *feature*. I want the v2 API to correctly communicate how to update records. That requires passing the update key.

Consequently, configuring a sink using `OutputMode.Update` is unsafe. It doesn't communicate the key. My position is (and has been) that we cannot use `OutputMode.Update` to configure v2 sinks because the update behavior -- which rows should be replaced -- is undefined.

Instead, I proposed a [mix-in to the `WriteBuilder`](https://docs.google.com/document/d/1vI26UEuDpVuOjWw4WPoH2T6y8WAekwtI7qoowhOFnI4/edit?disco=AAAACidsdOM):

```java
public interface SupportsUpdate extends WriteBuilder {
  WriteBuilder updateByKey(String... column)
}
```